File: internal/state/state_test.go
package state

import (
	"os"
	"path/filepath"
	"testing"
)

func TestLoadSaveRoundTrip(t *testing.T) {
	dir := t.TempDir()
	path := filepath.Join(dir, "state.json")

	snap := Snapshot{IP: "203.0.113.1", Ports: []string{"443", "22", "22"}}
	if err := Save(path, snap); err != nil {
		t.Fatal(err)
	}

	got, err := Load(path)
	if err != nil {
		t.Fatal(err)
	}
	if got.IP != snap.IP {
		t.Fatalf("IP = %q, want %q", got.IP, snap.IP)
	}
	if !PortsEqual(got.Ports, []string{"22", "443"}) {
		t.Fatalf("Ports = %v, want [22 443]", got.Ports)
	}
}

func TestLoadLegacyStateWithoutPorts(t *testing.T) {
	dir := t.TempDir()
	path := filepath.Join(dir, "state.json")
	if err := os.WriteFile(path, []byte(`{"last_ip":"203.0.113.1"}`+"\n"), 0o644); err != nil {
		t.Fatal(err)
	}

	got, err := Load(path)
	if err != nil {
		t.Fatal(err)
	}
	if got.IP != "203.0.113.1" {
		t.Fatalf("IP = %q", got.IP)
	}
	if len(got.Ports) != 0 {
		t.Fatalf("Ports = %v, want nil/empty", got.Ports)
	}
}

func TestPortsEqual(t *testing.T) {
	if !PortsEqual([]string{"22", "443"}, []string{"443", "22"}) {
		t.Fatal("expected equal ports")
	}
	if PortsEqual([]string{"22"}, []string{"22", "443"}) {
		t.Fatal("expected different ports")
	}
	if !PortsEqual(nil, []string{}) {
		t.Fatal("expected nil and empty to be equal")
	}
}

File: cpp/tests/state_test.cpp
#include "firewallkeeper/state/state.hpp"

#include <cassert>
#include <filesystem>
#include <fstream>
#include <iostream>

namespace fs = std::filesystem;

static int failures = 0;

#define CHECK(cond, msg)                         \
    do {                                         \
        if (!(cond)) {                           \
            std::cerr << "FAIL: " << msg << '\n'; \
            ++failures;                          \
        }                                        \
    } while (0)

int main() {
    auto dir = fs::temp_directory_path() / "fk_cpp_test";
    fs::create_directories(dir);
    auto path = (dir / "state.json").string();

    firewallkeeper::state::Snapshot snap;
    snap.ip = "203.0.113.1";
    snap.ports = {"443", "22", "22"};

    std::string err;
    CHECK(firewallkeeper::state::save(path, snap, err), "save failed: " + err);

    auto got = firewallkeeper::state::load(path, err);
    CHECK(err.empty(), "load error: " + err);
    CHECK(got.ip == snap.ip, "ip mismatch");
    CHECK(firewallkeeper::state::ports_equal(got.ports, {"22", "443"}), "ports mismatch");

    auto legacy_path = (dir / "legacy.json").string();
    {
        std::ofstream out(legacy_path);
        out << R"({"last_ip":"203.0.113.1"})" << '\n';
    }
    got = firewallkeeper::state::load(legacy_path, err);
    CHECK(got.ports.empty(), "legacy ports should be empty");
    CHECK(firewallkeeper::state::ports_equal({}, {}), "nil ports equal");

    fs::remove_all(dir);
    return failures == 0 ? 0 : 1;
}
